File: 2784-power-of-heroes/solution.ts

const mod = 10 ** 9 + 7;

function multiply(a: number, b: number) {
  if (a < b) return multiply(b, a);

  let res = 0;
  while (b) {
    if (b & 1) res = (res + a) % mod;
    a = (a * 2) % mod;
    b >>= 1;
  }
  return res;
}

function sumOfPower(nums: number[]): number {
  const len = nums.length;
  const powOf2 = [1];
  for (let i = 1; i <= len - 2; i++) {
    powOf2[i] = multiply(powOf2[i - 1], 2);
  }
  const multipliers: number[] = [];

  let answer = 0;
  nums.sort((a, b) => a - b);

  for (let i = 0; i < len; i++) {
    const square = multiply(nums[i], nums[i]);
    const multiplier =
      (nums[i] + multiply(multipliers[i - 1] ?? 0, 2) - (nums[i - 1] || 0)) %
      mod;
    multipliers[i] = multiplier;
    answer = (answer + multiply(square, multiplier)) % mod;
  }
  return answer;
}

File: 2784-power-of-heroes/solution.cpp
class Solution {
public:
    int mod = 1e9 + 7;
    int fastPow(long long a, int b) {
        long long res = 1;
        while(b > 0) {
            if(b & 1) res = res * a % mod;
            a = a * a % mod, b >>= 1;
        }
        return res;
    }
    
    int sumOfPower(vector<int>& nums) {
        sort(nums.rbegin(), nums.rend());
        int n = nums.size();
        long long ans = 0, base = 1, sSum = 0;
        for(int i = 1; i < n; i++) {
            sSum = (sSum + nums[i] * base) % mod;
            base = base * 2 % mod;
        }
        for(int i = 0; i < n; i++) {
            long long sq = (long long)nums[i] * nums[i] % mod;
            ans = (ans + sq * (nums[i] + sSum)) % mod;
            if(i + 1 < n) sSum = (sSum - nums[i + 1]) * fastPow(2, mod - 2) % mod;
        }
        return ans < 0 ? (ans + mod) : ans;
    }
};
